# 1. Bank Loan Modelling 

### Logistic Regression

In [1]:
import numpy as np
import pandas as pd

In [2]:
#import numpy as np
import statsmodels.api as sm

In [3]:
# loading the dataset
df = pd.read_excel('Bank_Personal_Loan_Modelling.xlsx', sheet_name=1)

In [4]:
df.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   ID                  5000 non-null   int64  
 1   Age                 5000 non-null   int64  
 2   Experience          5000 non-null   int64  
 3   Income              5000 non-null   int64  
 4   ZIP Code            5000 non-null   int64  
 5   Family              5000 non-null   int64  
 6   CCAvg               5000 non-null   float64
 7   Education           5000 non-null   int64  
 8   Mortgage            5000 non-null   int64  
 9   Personal Loan       5000 non-null   int64  
 10  Securities Account  5000 non-null   int64  
 11  CD Account          5000 non-null   int64  
 12  Online              5000 non-null   int64  
 13  CreditCard          5000 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 547.0 KB


> we do not have any missing values

In [6]:
df.shape

(5000, 14)

## Dependent variable 

In [7]:
y = df['Personal Loan']

In [8]:
df.columns

Index(['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Personal Loan', 'Securities Account',
       'CD Account', 'Online', 'CreditCard'],
      dtype='object')

## Independent variables

In [9]:
x= df[['Age', 'Experience', 'Income', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Securities Account',
       'CD Account', 'Online', 'CreditCard']]

## Adding the constant

In [10]:
x1 = sm.add_constant(x)

## Logit is the funcation for logistic regression

In [11]:
LR = sm.Logit(y,x1)

In [12]:
result =LR.fit()

Optimization terminated successfully.
         Current function value: 0.128435
         Iterations 9


In [13]:
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:          Personal Loan   No. Observations:                 5000
Model:                          Logit   Df Residuals:                     4988
Method:                           MLE   Df Model:                           11
Date:                Sat, 08 Aug 2020   Pseudo R-squ.:                  0.5938
Time:                        19:06:46   Log-Likelihood:                -642.18
converged:                       True   LL-Null:                       -1581.0
Covariance Type:            nonrobust   LLR p-value:                     0.000
======================================================================================
                         coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                -12.1928      1.645     -7.411      0.000     -15.417      -8.968
Age                   -0.0536      0.061     -0.874      0.382      -0.174       0.067
Experience             0.0638      0.061      1.046      0.295      -0.056       0.183
Income                 0.0546      0.003     20.831      0.000       0.049       0.060
Family                 0.6958      0.074      9.364      0.000       0.550       0.841
CCAvg                  0.1240      0.040      3.127      0.002       0.046       0.202
Education              1.7362      0.115     15.088      0.000       1.511       1.962
Mortgage               0.0005      0.001      0.856      0.392      -0.001       0.002
Securities Account    -0.9368      0.286     -3.277      0.001      -1.497      -0.377
CD Account             3.8225      0.324     11.800      0.000       3.188       4.457
Online                -0.6752      0.157     -4.298      0.000      -0.983      -0.367
CreditCard            -1.1197      0.205     -5.462      0.000      -1.522      -0.718
======================================================================================
"""

> As we can see P>|z| value of Income, Family, Education, CD Account, Online, CreaditCard is 0.00 ... Which indicates that these the most important features which decide the personal loan for the person i.e Yes/No.

> The features like CCAvg, Securities Account are also important for clasification as their P>|z| value is near to 0.00

> The other features like Age, experience, Mortgage are relatively less important overall.

# 2. Attrition Analysis 

### Logistic Regression

In [14]:
df2 = pd.read_csv('Attrition_anaysis.csv')
df2.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [15]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

In [16]:
df2['NumCompaniesWorked'].mean()

2.6948303347756775

In [17]:
df2['TotalWorkingYears'].mean()

11.279936378095888

In [18]:
new_NumCompaniesWorked  = np.where(df2['NumCompaniesWorked'].isnull(), 2.69, df2['NumCompaniesWorked'] )
df2['NumCompaniesWorked'] = new_NumCompaniesWorked

In [19]:
new_TotalWorkingYears  = np.where(df2['TotalWorkingYears'].isnull(), 11.27, df2['TotalWorkingYears'] )
df2['TotalWorkingYears'] = new_TotalWorkingYears

In [20]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 24 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Age                      4410 non-null   int64  
 1   Attrition                4410 non-null   object 
 2   BusinessTravel           4410 non-null   object 
 3   Department               4410 non-null   object 
 4   DistanceFromHome         4410 non-null   int64  
 5   Education                4410 non-null   int64  
 6   EducationField           4410 non-null   object 
 7   EmployeeCount            4410 non-null   int64  
 8   EmployeeID               4410 non-null   int64  
 9   Gender                   4410 non-null   object 
 10  JobLevel                 4410 non-null   int64  
 11  JobRole                  4410 non-null   object 
 12  MaritalStatus            4410 non-null   object 
 13  MonthlyIncome            4410 non-null   int64  
 14  NumCompaniesWorked      

In [21]:
from sklearn import preprocessing

In [22]:
le = preprocessing.LabelEncoder()

In [23]:
df2.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

In [24]:
df3 = df2.copy()

In [25]:
df3['Attrition'] = le.fit_transform(df2['Attrition'])
df3['BusinessTravel'] = le.fit_transform(df2['BusinessTravel'])
df3['Department'] = le.fit_transform(df2['Department'])
df3['EducationField'] = le.fit_transform(df2['EducationField'])
df3['Gender'] = le.fit_transform(df2['Gender'])
df3['JobRole'] = le.fit_transform(df2['JobRole'])
df3['MaritalStatus'] = le.fit_transform(df2['MaritalStatus'])
df3['Over18'] = le.fit_transform(df2['Over18'])

In [26]:
df3.head()

Age  Attrition  BusinessTravel  Department  DistanceFromHome  Education  \
0   51          0               2           2                 6          2   
1   31          1               1           1                10          1   
2   32          0               1           1                17          4   
3   38          0               0           1                 2          5   
4   32          0               2           1                10          1   

   EducationField  EmployeeCount  EmployeeID  Gender  ...  NumCompaniesWorked  \
0               1              1           1       0  ...                 1.0   
1               1              1           2       0  ...                 0.0   
2               4              1           3       1  ...                 1.0   
3               1              1           4       1  ...                 3.0   
4               3              1           5       1  ...                 4.0   

   Over18  PercentSalaryHike  StandardHours  StockOptionLevel  \
0       0                 11              8                 0   
1       0                 23              8                 1   
2       0                 15              8                 3   
3       0                 11              8                 3   
4       0                 12              8                 2   

   TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                1.0                      6               1   
1                6.0                      3               5   
2                5.0                      2               5   
3               13.0                      5               8   
4                9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

## Dependent variable 

In [27]:
y2 = df3['Attrition']

## Independent variables

In [28]:
x2= df3[['BusinessTravel', 'Department', 'EducationField', 'Gender', 'JobRole', 'MaritalStatus', 'Age',
         'DistanceFromHome', 'JobLevel', 'MonthlyIncome','NumCompaniesWorked', 'PercentSalaryHike',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'YearsAtCompany', 'YearsSinceLastPromotion', 
         'YearsWithCurrManager']]

## Adding the constant

In [29]:
import statsmodels.api as sm

In [30]:
x2c = sm.add_constant(x2)

## Logit is the funcation for logistic regression

In [31]:
LR = sm.Logit(y2,x2c)

In [32]:
result2 =LR.fit()

Optimization terminated successfully.
         Current function value: 0.393261
         Iterations 7


In [33]:
result2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Attrition   No. Observations:                 4410
Model:                          Logit   Df Residuals:                     4391
Method:                           MLE   Df Model:                           18
Date:                Sat, 08 Aug 2020   Pseudo R-squ.:                  0.1097
Time:                        19:06:49   Log-Likelihood:                -1734.3
converged:                       True   LL-Null:                       -1947.9
Covariance Type:            nonrobust   LLR p-value:                 1.931e-79
===========================================================================================
                              coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------------
const                      -0.1266      0.390     -0.325      0.745      -0.891       0.638
BusinessTravel             -0.0153      0.065     -0.234      0.815      -0.144       0.113
Department                 -0.2424      0.081     -2.984      0.003      -0.402      -0.083
EducationField             -0.0963      0.033     -2.888      0.004      -0.162      -0.031
Gender                      0.0861      0.089      0.962      0.336      -0.089       0.262
JobRole                     0.0367      0.018      2.055      0.040       0.002       0.072
MaritalStatus               0.5873      0.063      9.307      0.000       0.464       0.711
Age                        -0.0303      0.007     -4.450      0.000      -0.044      -0.017
DistanceFromHome           -0.0013      0.005     -0.234      0.815      -0.012       0.009
JobLevel                   -0.0273      0.040     -0.689      0.491      -0.105       0.050
MonthlyIncome           -1.854e-06   9.55e-07     -1.941      0.052   -3.73e-06    1.81e-08
NumCompaniesWorked          0.1160      0.018      6.327      0.000       0.080       0.152
PercentSalaryHike           0.0127      0.012      1.078      0.281      -0.010       0.036
StockOptionLevel           -0.0641      0.052     -1.243      0.214      -0.165       0.037
TotalWorkingYears          -0.0579      0.012     -4.853      0.000      -0.081      -0.034
TrainingTimesLastYear      -0.1472      0.035     -4.186      0.000      -0.216      -0.078
YearsAtCompany              0.0123      0.018      0.673      0.501      -0.024       0.048
YearsSinceLastPromotion     0.1314      0.020      6.420      0.000       0.091       0.172
YearsWithCurrManager       -0.1398      0.022     -6.315      0.000      -0.183      -0.096
===========================================================================================
"""

> As we can see P>|z| value of MaritalStatus, Age, YearsWithCurrManager, YearsSinceLastPromotion, TrainingTimesLastYear, NumCompaniesWorked	 is 0.00 ... Which indicates that these the most important features which decide the Attrition for the person i.e Yes/No.

> The features like Department, EducationField, JobRole, MonthlyIncome are also important for clasification as their P>|z| value is near to 0.00 and play important role in analysing the Attrition in company

> The other features like BusinessTravel, YearsAtCompany, StockOptionLevel, PercentSalaryHike, JobLevel, DistanceFromHome, Gender are relatively less important overall to anlysis the Attrition.

# 3. Real Estate Analysis 

### Linear Regression

In [34]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
dataset = pd.read_excel('Real_estate_analysis.xlsx', sheet_name=0)

In [35]:
dataset.head()

price  sqft_living  bedrooms  bathrooms  floors
0  221900         1180         3       1.00     1.0
1  538000         2570         3       2.25     2.0
2  180000          770         2       1.00     1.0
3  604000         1960         4       3.00     1.0
4  510000         1680         3       2.00     1.0

In [36]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   price        21613 non-null  int64  
 1   sqft_living  21613 non-null  int64  
 2   bedrooms     21613 non-null  int64  
 3   bathrooms    21613 non-null  float64
 4   floors       21613 non-null  float64
dtypes: float64(2), int64(3)
memory usage: 844.4 KB


## Model 1 
### Price v\s sqft

In [37]:
Y1 = dataset['price']
X1 = dataset[['sqft_living']]

X_train1, X_test1, Y_train1, Y_test1 = train_test_split(X1, Y1, test_size = 0.3, random_state=1)

model1 = LinearRegression()
model1.fit(X_train1, Y_train1)
print(model1.intercept_.round(2))
print(model1.coef_.round(2))

-31356.78
[273.81]


In [38]:
y_test_predicted1 = model1.predict(X_test1)
print(y_test_predicted1)

[ 732568.6468476   414950.90612649  617569.12003479 ...  319117.96711581
 1384232.63212023  858520.50954735]


In [39]:
mean_squared_error(Y_test1,y_test_predicted1).round(2)


77554859856.1

In [40]:
model1.score(X_test1, Y_test1)

0.503077461496144

## Model 2
### Price v\s bedrooms

In [41]:
Y2 = dataset['price']
X2 = dataset[['bedrooms']]

X_train2, X_test2, Y_train2, Y_test2 = train_test_split(X2, Y2, test_size = 0.3, random_state=1)

model2 = LinearRegression()
model2.fit(X_train2, Y_train2)
print(model2.intercept_.round(2))
print(model2.coef_.round(2))

152044.37
[114485.28]


In [42]:
y_test_predicted2 = model2.predict(X_test2)
print(y_test_predicted2)

[609985.5089382  495500.22465829 381014.94037837 ... 381014.94037837
 724470.79321811 609985.5089382 ]


In [43]:
mean_squared_error(Y_test2,y_test_predicted2).round(2)


140388576416.63

In [44]:
model2.score(X_test2, Y_test2)

0.1004787075970095

## Model 3
### Price v\s bathrooms

In [45]:
Y3 = dataset['price']
X3 = dataset[['bathrooms']]

X_train3, X_test3, Y_train3, Y_test3 = train_test_split(X3, Y3, test_size = 0.3, random_state=1)

model3 = LinearRegression()
model3.fit(X_train3, Y_train3)
print(model3.intercept_.round(2))
print(model3.coef_.round(2))

26476.47
[241566.37]


In [46]:
y_test_predicted3 = model3.predict(X_test3)
print(y_test_predicted3)

[690783.97703771 570000.79304754 388826.01706228 ... 268042.83307212
 871958.75302296 630392.38504262]


In [47]:
mean_squared_error(Y_test3,y_test_predicted3).round(2)


111707831778.36

In [48]:
model3.score(X_test3, Y_test3)

0.2842467971567646

## Model 4
### Price v\s floors

In [49]:
Y4 = dataset['price']
X4 = dataset[['floors']]

X_train4, X_test4, Y_train4, Y_test4 = train_test_split(X4, Y4, test_size = 0.3, random_state=1)

model4 = LinearRegression()
model4.fit(X_train4, Y_train4)
print(model4.intercept_.round(2))
print(model4.coef_.round(2))

285434.67
[168385.25]


In [50]:
y_test_predicted4 = model4.predict(X_test4)
print(y_test_predicted4)

[622205.17145496 453819.92254912 622205.17145496 ... 453819.92254912
 790590.4203608  453819.92254912]


In [51]:
mean_squared_error(Y_test4,y_test_predicted4).round(2)


145932995509.5

In [52]:
model4.score(X_test4, Y_test4)

0.06495357332084639

## Model 5
### Price v\s sqft, floors, bathrooms, bedrooms

In [53]:
Y5 = dataset['price']
X5 = dataset[['floors','bathrooms','bedrooms','sqft_living']]

X_train5, X_test5, Y_train5, Y_test5 = train_test_split(X5, Y5, test_size = 0.3, random_state=1)

model5 = LinearRegression()
model5.fit(X_train5, Y_train5)
print(model5.intercept_.round(2))
print(model5.coef_.round(2))

74783.06
[  3448.15   6088.37 -53543.38    300.84]


In [54]:
y_test_predicted5 = model5.predict(X_test5)
print(y_test_predicted5)

[ 723585.34725714  421664.94687757  696709.91578542 ...  362304.75392675
 1394049.5346314   857000.3321147 ]


In [55]:
mean_squared_error(Y_test5,y_test_predicted5).round(2)

75202494124.52

In [56]:
model5.score(X_test5, Y_test5)

0.5181499347492146

## Conclusion

- model 1 which is between price and sqft has compartivly high score of 50 %
- other model have score of around 28 %, 10 %, 6 %
- the 5th model which is between all teh 4 features and the price has score of 51.8%